core/backend/src/CodeService.ts

```diff
@@ -0,0 +1,521 @@
+    /** The Guid of the new code. This must be always be supplied by the application. */
+    readonly guid: CodeGuid;
+    /** An optional value for the state of the code. */
+    readonly state?: CodeState;
+    /** An optional json object to be stored with the code. */
+    readonly json?: SettingObject;
+  }
+
+  /** Properties of a "proposed" new code to be reserved.
+   * @note the Guid of the entity identified by this code *must* be supplied, but `value` is optional, since
+   * this may be used to reserve codes from a sequence where the value is generated.
+   */
+  export interface ProposedCodeProps extends CodeGuidStateJson {
+    /** The value for the proposed code.
+     * @note For code sequence operations, this value is ignored on input and is set with a new value from the sequence on successful return.
+     */
+    value?: CodeValue;
+  }
+
+  /** Properties of a proposed new code that is not from a code sequence (its `value` is required). */
+  export type ProposedCode = ProposedCodeProps & ScopeSpecAndValue;
+
+  /** Properties that describe a code sequence and a scope and spec for a proposed code or array of codes from a sequence. */
+  export interface SequenceScope extends ScopeAndSpec {
+    /** The code sequence. */
+    readonly seq: CodeSequence;
+    /** A valid current value. If supplied, the returned value will always be later in the sequence than this. */
+    readonly start?: CodeValue;
+  }
+
+  /** Properties of a code to be updated.
+   * @note The `guid` member identifies the code to be updated and is required.
+   * All other properties are optional - if `undefined`, its value is not changed.
+   */
+  export type UpdatedCode = MarkRequired<Partial<ProposedCode>, "guid">;
+
+  /** A proposed code that could not be reserved due to some error. */
+  export interface ReserveProblem {
+    /** the proposed code that failed. */
+    readonly code: ProposedCode;
+    /** the reason for the failure */
+    readonly errorId: ErrorId;
+    /** the error message from the exception for this proposed code. */
+    readonly message: string;
+  }
+
+  /** A update to a code that failed for some error. */
+  export interface UpdateProblem {
+    /** The properties of the code that was to be updated */
+    readonly prop: UpdatedCode;
+    /** the reason for the failure */
+    readonly errorId: ErrorId;
+    /** the error message from the exception for the update request. */
+    readonly message: string;
+  }
+
+  /**
+   * A sequence of code values following a increasing pattern. Valid code sequences must have a first value, a last value, and
+   * a way to get the next valid value from an existing valid value.
+   * Code sequences have a `sequenceName` so they can be registered using `CodeService.registerSequence`.
+   */
+  export interface CodeSequence {
+    /** the name of this CodeSequence. */
+    get sequenceName(): string;
+    /** the type of this CodeSequence. Usually this is the class name of the sequence. */
```

Comment:
"class name" meaning "BIS class name" (ViewDefinition, Category, etc)?

Reply:
There really isn't a use for this yet. I just removed it.